src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -308,10 +308,10 @@ class HandshakeALotClosure : public HandshakeClosure {
   }
 };
 
-bool VMThread::handshake_alot() {
+bool VMThread::handshake_or_safepoint_alot() {
   assert(_cur_vm_operation == nullptr, "should not have an op yet");
   assert(_next_vm_operation == nullptr, "should not have an op yet");
-  if (!HandshakeALot) {
+  if (!HandshakeALot && !SafepointALot) {
     return false;
   }
   static jlong last_halot_ms = 0;
```

Comment:
Below this point is this code:
```
  // If only HandshakeALot is set, but GuaranteedSafepointInterval is 0,
  // we emit a handshake if it's been more than a second since the last one.
  jlong interval = GuaranteedSafepointInterval != 0 ? GuaranteedSafepointInterval : 1000;
```

Do you need this now that you've made the change in src/hotspot/share/runtime/arguments.cpp?

Reply:
You're right.  This has strange logic.  We only set GuaranteedSafepointInterval in arguments.cpp if the flag hasn't been set on the command line.  So someone could set explicitly set GuaranteeSafepointInterval=0.   If GuaranteedSafepointInterval is 0 then we can get to this code if another safepoint operation wakes up the VM Thread.  And in this case, we want to make sure it's been longer than a second ago. 

I fixed the comment and changed the variable name to alot_ms.